### ISSUE TITLE
How to set parameters properly to accelerate calculation?

### ISSUE THREAD

**Issue by chen314156** [shankar1729/jdftx#43]
Hi dear shankar，
I‘m a beginner and I want to optimize N-doped graphene with JDFTx, I used this command:
include NiN4.ionpos
include NiN4.lattice
ion-species GBRV/$ID_pbe.uspp
elec-cutoff 20 100
coulomb-interaction Slab 001
elec-ex-corr gga-PBE
coulomb-truncation-embed 0 0 0.11332
coords-type Lattice
ionic-minimize \
    nIterations 10 \
    energyDiffThreshold 1e-6 \
    knormThreshold 2e-4
kpoint-folding 3 3 1
elec-smearing Fermi 0.01
electronic-SCF
add-U Ni d 3.79
fluid LinearPCM
pcm-variant CANDLE
fluid-solvent H2O
dump-name NiN4.$VAR
dump Ionic IonicPositions ElecDensity BoundCharge

I want to know is there something wrong with my input? The test takes much time and it's still unfinished.
 Thank you very much.

**Comment by shankar1729** [comment]
How are you running this input file, and what is your hardware? Its difficult to determine performance reasons without that information. Please also attach the ionpos, lattice and input files; pasting them as text is removing some key characters like the line continuation.

Best,
Shankar

**Comment by shankar1729** [comment]
Hi Chen,

It seems that the issue is due to using 48 cores all in MPI, rather than in MPI-threads mode  as JDFTx requires for performance. See http://jdftx.org/GettingStarted.html for more information about this.

In particular for your system, you have nStates = 5 (reduced kpoints). JDFTx only parallelizes these over MPI. So you should use five MPI processes and many threads for each. Given that you have 24 cores per node, it would be ideal to run on 2.5 or 5 nodes using either "mpirun -n 5 jdftx -c12 ..." or "mpirun -n 5 jdftx -c24 ..". In either case, make sure you look into cpu binding of your MPI and use --bind-to none or --bind-to socket as appropriate. (Else you may get all threads locked onto a single core.)

Very likely, once you fix these issues you can speed up your calculations by more than an order of magnitude.

Best,
Shankar

**Comment by chen314156** [comment]
Hi, Shankar,
I have done the k-point test and find that 2 2 1 can be the best set. However,  when I set the same parameters for NiN3,  it always reminds me in the out files as follows : 
IonicMinimize: None of the convergence criteria satisfied after 10 iterations

My NiN3.in:
include NiN3.ionpos
include NiN3.lattice
ion-species GBRV/$ID_pbe.uspp
elec-cutoff 20 100
coulomb-interaction Slab 001
elec-ex-corr gga-PBE
coulomb-truncation-embed 0 0 0.14003
coords-type Lattice
ionic-minimize \
    nIterations 10 \
    energyDiffThreshold 1e-6 \
    knormThreshold 2e-4
kpoint-folding 2 2 1
elec-smearing Fermi 0.01
electronic-SCF
add-U Ni d 3.79
fluid LinearPCM
pcm-variant CANDLE
fluid-solvent H2O
dump-name NiN3.$VAR
dump Ionic IonicPositions ElecDensity BoundCharge

Should I change nIterations to 100 or another? or change another parameters to decrease the convergence criteria? 

You recommended to fix the issues to speed up the calculation, but I want to get the optimal positions of the issues for the subsequent calculations.

Best wishes,
Chen

**Comment by shankar1729** [comment]
Yes, you might need to increase the iterations of ionic minimize.

Best,
Shankar

**Comment by Rachel-wang2** [comment]
Hi chen, i also meet the same problems just like you, the Ionminimize step is 100,but the process of calculation is slow and always not finished. i want to know wheather your problem is solved or not?

**Comment by chen314156** [comment]
> Hi chen, i also meet the same problems just like you, the Ionminimize step is 100,but the process of calculation is slow and always not finished. i want to know wheather your problem is solved or not?

Hi Rachel-wang, when I set the Ionminimize step to 100, the calculation process is slow and not finished. I tried to enlarge  knormThreshold, it converges quickly, but for the other calculation system, it was slow. Maybe you can try to enlarge knormThreshold and energyDiffThreshold, spontaneously, I'm try it too.

**Comment by chen314156** [comment]
> Hi chen, i also meet the same problems just like you, the Ionminimize step is 100,but the process of calculation is slow and always not finished. i want to know wheather your problem is solved or not?

Hi Rachel-wang,  did you meet the problem, the process created too large output files .lsbatch (about 2.4T)? and it also created large .log files (about 100 G) when being killed.

**Comment by shankar1729** [comment]
Hi Chen,

Something is probably off with your build if it's producing that large of an output file. If you tail these output files, do you see a whole bunch of stack traces in an infinite loop?

What operating system and compiler version are you using? Can you post the CMakeCache.txt from your build directory?

Best,
Shankar

**Comment by chen314156** [comment]
Hi Shankar,
The output files do have a whole bunch of stack traces in an infinite loop. My operating system  is CentOS 7.5 Linux x86_64 and compiler version is GCC. This is the CMakeCache.txt file from my build directory below.
[CMakeCache.txt](https://github.com/shankar1729/jdftx/files/2552829/CMakeCache.txt)

**Comment by shankar1729** [comment]
You seem to be using the Intel mpi compiler but system gcc for the regular compiler. This mismatch could produce an issue.

Also check what was the last output before the stack traces begin.

Finally, for performance, you may want to link against  MKL if you have it.

Best,
Shankar

**Comment by chen314156** [comment]
Dear Shankar

Thanks for your reply. Intel mpi compiler doesn't match gcc. Can I just use intel mpi compiler to build jdftx? If intel mpi compiler doesn't work, which compiler I can use to build jdftx for paralleling the job. Does MKL means Intel MKL?

Best regards,
Chen

**Comment by shankar1729** [comment]
Yes, intel compiler should work. But make sure the intel compiler uses a new enough gcc/glibc under the hood.

Alternately, you can use open-mpi or mpich built with the same compiler as your system gcc. (My cluster uses centos 7.5 too and we use open-mpi built with the system gcc 4.8.)

Yes, MKL is Intel MKL. You can use that regardless of whether you use intel or gcc compilers.

Best,
Shankar

**Comment by Rachel-wang2** [comment]
Hi chen, I haven’t meet this problem for now, it may be your input file of common.in  include some extra dump option.Usually  what kind of file  should be write  as output , I  set inputfile  following  the tutorial of JDFTx.

发送自 Windows 10 版邮件应用

发件人: chen314156
发送时间: 2018年11月6日 16:09
收件人: shankar1729/jdftx
抄送: Rachel-wang2; Comment
主题: Re: [shankar1729/jdftx] How to set parameters properly to acceleratecalculation? (#43)

Hi chen, i also meet the same problems just like you, the Ionminimize step is 100,but the process of calculation is slow and always not finished. i want to know wheather your problem is solved or not?
Hi Rachel-wang, did you meet the problem, the process created too large output files .lsbatch (about 2.4T)? and it also created large .log files (about 100 G) when being killed.

**Comment by chen314156** [comment]
Dear Shankar,
The calculation process is so slow and always not finished. I want to know if it is possible to use the structure optimized by VASP to continue the calculation in JDFTx, eg. the free energy.

**Comment by shankar1729** [comment]
Yes, you can do that.

Were you able to fix the previous issues?

**Comment by chen314156** [comment]
> Yes, you can do that.
> 
> Were you able to fix the previous issues?
Hi Shankar,
Yes, I recompiled JDFTx, and it calculated successfully without creating large output files.  Thank you very much.

Shankar, I also want to ask if it is still ok to use the structure calculated by VASP when I set potential value in the calculation system?
Regards,
Chen

**Comment by Sparkliang** [comment]
Hi dear shankar，
 I wanted to optimize and calculate DOS of N-doped graphene with JDFTx, and the supercell was 6*6 with 71 atoms in hexagonal lattice of a=27.94(Bohr) and c=38(Bohr). In the relax optimization step, the kpoint-folding was 3 3 1, and I used 4 nodes 112 cores (28 cores per node). It took around 20 hours to finish the relax optimization. In the DOS calculation, the kpoint-folding was 39 39 1. When I used 1 node 28 cores, the calculation always stopped at “Allocating electronic variables” and created a jdftx-stacktrace file. I have attached related files. So in my case, I have following confusing questions:
1.	The relax optimization of 71 atoms in 112 cores took around 20 hours. Is it normal? Were all threads locked onto a single core? What should be the normal time of this condition?
2.	In the DOS calculation with kpoint-folding of 39 39 1, what’s the reason of stopping at “Allocating electronic variables” in calculation with 1 node 28 cores.
3.	I found that in all JDFTx web examples, at least one ion is put at (0 0 0) position, or one layer of ions at z=0 plane for 2D materials. however, in VASP software, ions are usually not put at the boundary of lattice. So I wonder the reason of put ions at the boundary in JDFTx.
I would be very appreciated for your answer.
Best regards!
[N-doped-gra.zip](https://github.com/shankar1729/jdftx/files/2612254/N-doped-gra.zip)

**Comment by chen314156** [comment]
Hi dear Shankar,
I have one more question to bother you. I'm calculating the free energy for the H atom on N-doped graphene with JDFTx. According to your manual, I kept the surface fixed, and only set H atom to ''1" in .ionpos file. I used two nodes 48 cores, the calculation always reminds me in the .out files as follows:

Linear fluid (dielectric constant: 78.4) occupying 0.808179 of unit cell:       Completed after 0 iterations at t[s]:  36871.44
ElecMinimize:   Step increased Etot by 3.358929e-08, reducing alpha to 1.542727e-04.
ElecMinimize:   Step failed to reduce Etot after 3 attempts. Quitting step.
ElecMinimize:   Undoing step.

However, the calculation process is still calculating without stopping. I was wondering if it is wrong with my calcaltion. My input file is ：
include test.ionpos
include test.lattice
ion-species GBRV/$ID_pbe.uspp
elec-cutoff 20 100
coulomb-interaction Slab 001
elec-ex-corr gga-PBE
coulomb-truncation-embed 0 0 0.5
fluid LinearPCM
pcm-variant CANDLE
fluid-solvent H2O
dump-name test.$VAR
dump End None
vibrations \
    dr 0.01 \
    centralDiff yes \
        T 298

Best regards,
Chen

**Comment by shankar1729** [comment]
Hi @chen314156,

The failed step can occur one in a while, and the minimizer is written to be able to recover from that. So I wouldn't flag an error based on that alone. But if you send me the whole output, I can look in to see if it looks reasonable.

Best,
Shankar

**Comment by chen314156** [comment]
Hi Shankar,
Thank you for your reply. This is my output files below:

[test.zip](https://github.com/shankar1729/jdftx/files/2617886/test.zip)

Regards, 
Chen

**Comment by shankar1729** [comment]
Hi Chen,

That output seems mostly reasonable. The issue seems to be that to reach electronic convergence at the 1e-8 level, the fluid minimize in the inner step is not converging sufficiently. So try adjusting the convergence of the fluid-minimize using "fluid-minimize knormThreshold 3e-12" (reduced from 1e-11 default).

Also, your run configuration is far from optimal, which would probably explain the slowness. You are using 48 processes with 1 core each for a job with a single reduced k-point (nStates=1). Instead, run a single MPI process on one node with all 24 cores allocated to it. If your job management system is SLURM, you would do this with -n 1 -c 24 specified to sbatch / srun (search for equivalents if you are using Torque/PBS). If your MPI is OpenMPI, also specify "--bind-to none" in mpirun within your job file, or figure out appropriate cpu affinity commands for your mpi / job manager. 

JDFTx is a hybrid MPI-threads implementation, and the incorrect job configuration is the reason for your poor reported performance. I suspect you should see an order of magnitude performance boost once you fix these issues. Please let me know if that works for you.

Best,
Shankar

**Comment by chen314156** [comment]
Hi Shankar,
I deeply appreciate your help. Your former suggestion worked well with my calculation system (only contains 9 atoms). However, for my another calculated system (72 atoms), it always lasts too much time.  I'm not sure what on earth the problem is. It always upsets me.  I just want to calculate the free energy of the adsorbed species. My test files are below. Thank you again.
[1NH.zip](https://github.com/shankar1729/jdftx/files/2635965/1NH.zip)
Best regards, 
Chen
